examples/usercall-extension-bind/runner/src/main.rs

```diff
@@ -0,0 +1,356 @@
+/// This can be useful to strip protocol encapsulations, say while servicing requests load balanced by HA Proxy.
+/// This example demonstrates de-encapsulation for various HA proxy configurations before handing over the stream to the enclave.
+/// To simulate HA proxy configurations, the runner spawns a thread that connects to the same address which enclave binds to and 
+/// writes encapsulated test data for various HA proxy configurations to the stream.
+
+
+const SIZE_HEADER_SIG: usize = 12;
+static HEADER_SIG: [u8; SIZE_HEADER_SIG] = [0x0d, 0x0a, 0x0d, 0x0a, 0x00, 0x0d, 0x0a, 0x51, 0x55, 0x49, 0x54, 0x0a];
+const V2CMD_LOCAL: u8 = 0x2_0;
+const V2CMD_PROXY: u8 = 0x2_1;
+const FAMILY_UNSPEC: u8 = 0;
+const FAMILY_TCP4: u8 = 0x11;
+const FAMILY_TCP6: u8 = 0x21;
+
+
+struct ProxyIpV4Addr {
+    src_addr: u32,
+    dst_addr: u32,
+    src_port: u16,
+    dst_port: u16,
+}
+
+impl ProxyIpV4Addr {
+    fn from_reader(rdr: &mut impl Read) -> IoResult<Self> {
+        let src_addr = rdr.read_u32::<NetworkEndian>()?;
+        let dst_addr = rdr.read_u32::<NetworkEndian>()?;
+        let src_port = rdr.read_u16::<NetworkEndian>()?;
+        let dst_port = rdr.read_u16::<NetworkEndian>()?;
+
+        Ok(ProxyIpV4Addr {
+            src_addr,
+            dst_addr,
+            src_port,
+            dst_port,
+        })
+    }
+}
+impl From<ProxyIpV4Addr> for (SocketAddr, SocketAddr) {
+    fn from(proxy_addr: ProxyIpV4Addr) -> (SocketAddr, SocketAddr) {
+        let src_ip_addr : Ipv4Addr = proxy_addr.src_addr.into();
+        let dst_ip_addr : Ipv4Addr = proxy_addr.dst_addr.into();
+        let local_addr = SocketAddr::new(src_ip_addr.into(), proxy_addr.src_port);
+        let peer_addr = SocketAddr::new(dst_ip_addr.into(), proxy_addr.dst_port);
+        (local_addr, peer_addr)
+    }
+}
+
+#[derive(Default)]
+struct ProxyIpV6Addr {
+    src_addr: [u8; 16],
+    dst_addr: [u8; 16],
+    src_port: u16,
+    dst_port: u16,
+}
+
+impl ProxyIpV6Addr {
+    fn from_reader(rdr: &mut impl Read) -> IoResult<Self> {
+        let mut addr = ProxyIpV6Addr::default(); 
+
+        let _ = rdr.read_exact(&mut addr.src_addr[..])?;
+        let _ = rdr.read_exact(&mut addr.dst_addr[..])?;
+        addr.src_port = rdr.read_u16::<NetworkEndian>()?;
+        addr.dst_port = rdr.read_u16::<NetworkEndian>()?;
+
+        Ok(addr)
+    }
+}
+
+impl From<ProxyIpV6Addr> for (SocketAddr, SocketAddr) {
+    fn from(proxy_addr: ProxyIpV6Addr) -> (SocketAddr, SocketAddr) {
+        let src_ip_addr : Ipv6Addr = proxy_addr.src_addr.into();
+        let dst_ip_addr : Ipv6Addr = proxy_addr.dst_addr.into();
+        let local_addr = SocketAddr::new(src_ip_addr.into(), proxy_addr.src_port);
+        let peer_addr = SocketAddr::new(dst_ip_addr.into(), proxy_addr.dst_port);
+        (local_addr, peer_addr)
+    }
+}
+
+enum ProxyAddrReader {
+    V4(u16),
+    V6(u16),
+    UNSPEC(u16),
+}
```

Comment:
I think this is better since you always have a length:
```rust
enum ProxyAddrType {
    V4,
    V6,
    Unspec,
}

struct ProxyAddrReader {
    ty: ProxyAddrType,
    len: u16,
}
```

Reply:
Incorporated in recent push.